packages/big-design/src/components/Tree/reducer.utils.ts

```diff
@@ -0,0 +1,181 @@
+import { Action } from './reducer';
+import { MapValues, TreeNodeId, TreeNodeProps, TreeProps, TreeState } from './types';
+
+export const initialize = <T>(nodes: TreeProps<T>['nodes'], radio: boolean): TreeState<T> => {
+  const initialState: TreeState<T> = {
+    nodeMap: new Map<TreeNodeId, MapValues>(),
+    expandedNodeIds: new Set(),
+    flattenedNodeIds: [],
+    focusedNode: null,
+    selectedValues: new Set(),
+    visibleNodeIds: [],
+  };
+
+  return recursiveInitialize(nodes, initialState, radio);
+};
+
+const recursiveInitialize = <T>(
+  nodes: TreeProps<T>['nodes'] | undefined,
+  state: TreeState<T>,
+  radio: boolean,
+): TreeState<T> => {
+  if (!nodes || nodes.length < 1) {
+    return state;
+  }
+
+  // Order of statements in this loop is important.
+  nodes.forEach((node) => {
+    state.flattenedNodeIds.push(node.id);
+
+    state.nodeMap.set(node.id, {
+      children: new Set(node.children?.map((child) => child.id)),
+      id: node.id,
+      parent: getParentId(state.nodeMap, node.id),
+    });
+
+    if (state.focusedNode === null) {
+      state.focusedNode = node.id;
```

Comment:
I think the same comment applies here. I think `state` is an object that gets passed in from a reducer. If we mutate the object, it'd be much harder for the caller to detect changes.

Reply:
`focusedNode` is a value, not a reference so I think it would be fine in this case.